text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1563 @@
+With this RFC we extend the pattern syntax to allow type ascription inside of
+patterns. What this means is that `MyPattern : Type` is itself a valid pattern.
+For example, you may write:
+
+```rust
+match compute_stuff() {
+    Ok(vec: Vec<u8>) => {
+        // Logic...
+    },
+    Err(err: MyError<Foo>) => {
+        // Logic...
+    },
+}
+```
+
+The following is also valid:
+
+```rust
+match do_stuff() {
+    None => ...,
+    // We don't recommend this way of writing but it is possible:
+    Some(x): Option<u8> => ...,
+}
+
+if let Thing { field: binding: MyType } = make_thing() {
+    ...
+}
+```
+
+### In macros
+
+Just as we noted before that type ascription work in expression macros so may
+you use type ascription in pattern macros. For example:
+
+```rust
+macro_rules! ascribe {
+    ($p: pat, $n: expr) => {
+        $p : [u8, $n]
+    }
+}
+
+fn main() {
+    let ascribe!([x, y, z], 3) = [3, 1, 2];
+}
+```
+
+### Let bindings
+
+Before this RFC when you wrote something like:
+
+```rust
+let quux: u8 = 42;
+    ^^^^
+```
+
+The underlined part was the pattern, but the typing annotation `: u8` to
+the right was *not* part of the pattern. With this RFC, we unify the language
+and we can now say that everything after `let` and before `=` is the pattern:
+
+```rust
+let quux: u8 = 42;
+    ^^^^^^^^
+    Pattern!
+```
+
+Another implication of introducing type ascription in pattern contexts is that
+that you may say things like:
+
+```rust
+let [alpha: u8, beta, gamma] = [1, 2, 3];
+
+let (alpha: u8, beta: i16, gamma: bool) = (1, -2, true);
+```
+
+### Function definitions
+
+Another change that comes with this RFC is that the fundamental unit of a
+function parameter also becomes a pattern. Thus when you write things like:
+
+```rust
+fn foo(alpha: usize, beta: bool) { ... }
+       ^^^^^^^^^^^^  ^^^^^^^^^^^
+       Pattern       Pattern
+```
+
+the underlined parts are the patterns while before this RFC, the patterns were:
+
+```rust
+fn foo(alpha: usize, beta: bool) { ... }
+       ^^^^^         ^^^^
+       Pattern       Pattern
+```
+
+Indeed, with the change to function definitions, it becomes
+*syntactically valid* to write:
+
+```rust
+fn foo(alpha, beta) -> usize { .. }
+```
+
+Nevertheless, as we want to avoid introducing global type inference to the
+language, the type checker will prevent this from compiling and will emit an
+error:
+
+```rust
+error[E0282]: type annotations needed
+ --> src/main.rs:1:9
+  |
+1 |     fn foo(alpha, beta) -> usize { .. }
+  |            ^^^^^  ^^^^
+  |            |      |
+  |            |______|
+  |            |
+  |            The following patterns do not have a fully determined type.
+  |            help: Write type annotations on the following patterns:
+  |
+  |                alpha: usize
+  |                beta: bool
+  |
+
+error: aborting due to previous error
+```
+
+This also gives the compiler an opportunity to tell you what the types are
+if it so happens that you need this help.
```

Comment:
I think this would be hard right now unless it's a trait impl, so you might want to have the code example explicitly be such, or change the error to just say that you need types, without suggesting them.  (Especially since diagnostic quality is something that doesn't need RFCing -- people will just make it better even if the RFC doesn't say so, in the cases that are feasible.)

Reply:
Nuanced the language a bit here to make it optional :)